elf_binfmt.c

```diff
@@ -0,0 +1,98 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2024, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <uk/assert.h>
+#include <uk/arch/ctx.h>
+#include <uk/binfmt.h>
+#include <uk/essentials.h>
+#include <uk/errptr.h>
+#include <uk/init.h>
+
+#if CONFIG_LIBUKRANDOM
+#include <uk/random.h>
+#endif /* CONFIG_LIBUKRANDOM */
+
+#include "elf_prog.h"
+
+static int uk_binfmt_load_elf(struct uk_binfmt_loader_args *args)
+{
+	struct elf_prog *prog;
+	__u64 rand[2];
+	int rc;
+
+	UK_ASSERT(args);
+	UK_ASSERT(args->alloc);
+
+	prog = elf_load_vfs(args->alloc, args->pathname, args->progname);
+	if (unlikely(PTRISERR(prog))) {
+		rc = PTR2ERR(prog);
+		if (rc == -ENOEXEC) {
+			uk_pr_warn("%s not handled by ELF binfmt loader\n",
+				   args->pathname);
+			return UK_BINFMT_NOT_HANDLED;
+		}
+		uk_pr_err("Could not load ELF (%d)\n", rc);
+		return rc;
+	}
+
+	/* Save to private data in case we are requested to unload */
+	args->user = (void *)prog;
+
+	uk_pr_debug("%s: ELF loaded to 0x%lx-0x%lx (%lx B)\n", args->progname,
+		    (__u64)prog->vabase, (__u64)prog->vabase + prog->valen,
+		    prog->valen);
+	uk_pr_debug("%s: Entry at %p\n", args->progname, (void *)prog->entry);
+
+#if CONFIG_LIBUKRANDOM
+	uk_random_fill_buffer(rand, sizeof(rand));
+#else /* !CONFIG_LIBUKRANDOM */
+	/* Without random numbers, use a hardcoded seed */
+	uk_pr_warn("%s: Using hard-coded random seed\n", args->progname);
+	rand[0] = 0xB0B0;
+	rand[1] = 0xF00D;
+#endif /* !CONFIG_LIBUKRANDOM */
+
+	rc = elf_arg_env_count(&args->argc, args->argv,
+			       &args->envc, args->envp,
+			       args->stack_size);
+	if (unlikely(rc < 0)) {
+		uk_pr_err("Could not load ELF (%d)\n", rc);
+		elf_unload(prog);
```

Comment:
I feel like this is a bit awkward. Maybe we could first call `elf_arg_env_count` and then `elf_load_vfs`? Just so we avoid unloading on the error path. Or is there something stopping us from doing that?

Reply:
That's true, and I'm not very fond of it either, but the problem is that before interpreting the parameters the loader needs to determine first whether it's the right loader for this type of file. Right now this is done by `elf_vfs_load()` which is quite monolithic, so doing it right would require some refactoring there. I'll add a `TODO`.